cdc/cdc.cc

```diff
@@ -0,0 +1,471 @@
+        for (const auto& column : columns) {
+            b.with_column("_" + column.name(), column.type);
+        }
+    };
+    add_columns(s.partition_key_columns());
+    add_columns(s.clustering_key_columns());
+    add_columns(s.static_columns());
+    add_columns(s.regular_columns());
+    return ctx._migration_manager.announce_new_column_family(b.build(), false);
+}
+
+static future<> setup_desc(db_context ctx, const schema& s) {
+    schema_builder b(s.ks_name(), desc_name(s.cf_name()));
+    b.set_comment(sprint("CDC description for %s.%s", s.ks_name(), s.cf_name()));
+    b.with_column("node_ip", inet_addr_type, column_kind::partition_key);
+    b.with_column("shard_id", int32_type, column_kind::partition_key);
+    b.with_column("created_at", timestamp_type, column_kind::clustering_key);
+    b.with_column("stream_id", uuid_type);
+    return ctx._migration_manager.announce_new_column_family(b.build(), false);
+}
+
+template <typename Func>
+static void for_each_endpoint(db_context ctx, Func&& f) {
+    for (const auto& dc : ctx._token_metadata.get_topology().get_datacenter_endpoints()) {
+        for (const auto& endpoint : dc.second) {
+            f(endpoint);
+        }
+    }
+}
+
+template <typename Func>
+static void
+for_each_shard(db_context ctx, const gms::inet_address& endpoint, Func&& f) {
+    const unsigned int shard_count = ctx._snitch->get_shard_count(endpoint);
+    for (unsigned int shard_id = 0; shard_id < shard_count; ++shard_id) {
+        f(shard_id, shard_count);
+    }
+}
+
+// This function assumes setup_desc was called on |s| before the call to this
+// function.
+static future<> populate_desc(db_context ctx, const schema& s) {
+    auto& db = ctx._proxy.get_db().local();
+    auto desc_schema =
+        db.find_schema(s.ks_name(), desc_name(s.cf_name()));
+    auto log_schema =
+        db.find_schema(s.ks_name(), log_name(s.cf_name()));
+    auto& replication_strategy =
+        db.find_keyspace(s.ks_name()).get_replication_strategy();
+    // The assumption here is that all nodes in the cluster are using the
+    // same murmur3_partitioner_ignore_msb_bits. If it's not true, we will
+    // have to gossip it around together with shard_count.
+    auto ignore_msb_bits =
+        db.get_config().murmur3_partitioner_ignore_msb_bits();
+
+    auto belongs_to = [&](const gms::inet_address& endpoint,
+                          const unsigned int shard_id,
+                          const int shard_count,
+                          const utils::UUID& stream_id) {
+        const auto log_pk = partition_key::from_singular(*log_schema,
+                                                         data_value(stream_id));
+        const auto token = ctx._partitioner.decorate_key(*log_schema, log_pk).token();
+        auto&& eps =
+            replication_strategy.calculate_natural_endpoints(token, ctx._token_metadata);
+        if (eps.empty() || eps[0] != endpoint) {
+            return false;
+        }
```

Comment:
Unfortunately I found out that this approach is not correct.

If we use NetworkTopologyStrategy and the node we're trying to generate `stream_id` for belongs to a datacenter with RF = 0, the algorithm will never end, because the node will not be a natural endpoint for any token (in other words, no token range is primary for this node; this node is not a replica of the vnodes assigned to it).

What we should be looking for instead is a `stream_id` that belongs to a token range that is _owned_ by this node:
```
ctx._token_metadata.get_endpoint(ctx._token_metadata.first_token(token)) == endpoint
```
Then mutations with this `stream_id` will go exactly to the nodes that are replicas for the token ranges (vnodes) owned by this node. And I think that this is our intention.

Reply:
Will fix. Good catch.

Reply:
Or we could change the definition of the desc table and say that only nodes that are primary replicas (for some token ranges) have corresponding entries in the table, but then:
1. that would probably be harder to implement and introduce more complications when topology changes are handled
2. that would require an additional assumption: the set of replicas for a token range is determined by the primary replica. I think for our current implementations of replication strategies that is true, but I don't know if it's by design or by accident.
And the situation where the owner of a range is not its primary replica is rare, so we shoud optimize for the common case I guess.